src/team_lib/multirail/xccl_mrail_lib.c

```diff
@@ -0,0 +1,460 @@
+        if (status != XCCL_OK) {
+            xccl_mrail_debug("Failed to init internal teams");
+            goto free_teams;
+        }
+    }
+
+    XCCL_TEAM_SUPER_INIT(team->super, context, params);
+
+    xccl_mrail_debug("Exiting team create");
+    *tl_team = &team->super;
+    return XCCL_OK;
+free_teams:
+    team->n_teams -= 1;
+    for(; team->n_teams >= 0; team->n_teams--) {
+        xccl_team_destroy(team->teams[team->n_teams]);
+    }
+    free(team);
+    return status;
+}
+
+xccl_status_t xccl_mrail_team_create_test(xccl_tl_team_t *tl_team)
+{
+    xccl_mrail_team_t *team = ucs_derived_of(tl_team, xccl_mrail_team_t);
+    xccl_status_t     status;
+    int               i;
+
+    for(i = 0; i < team->n_teams; i++) {
+        status = xccl_team_create_test(team->teams[i]);
+        if (status == XCCL_INPROGRESS) {
+            return status;
+        }
+
+        if (status != XCCL_OK) {
+            xccl_mrail_error("internal team create failed");
+            goto free_teams;
+        }
+    }
+
+    return XCCL_OK;
+free_teams:
+    team->n_teams -= 1;
+    for(; team->n_teams >= 0; team->n_teams--) {
+        xccl_team_destroy(team->teams[team->n_teams]);
+    }
+    free(team);
+}
+
+xccl_status_t xccl_mrail_team_destroy(xccl_tl_team_t *team)
+{
+    xccl_mrail_team_t *tl = ucs_derived_of(team, xccl_mrail_team_t);
+
+    xccl_mrail_debug("Entering team destroy");
+    free(tl);
+    xccl_mrail_debug("Exiting team destroy");
+
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_mrail_context_progress(xccl_tl_context_t *team_context)
+{
+
+    return XCCL_OK;
+}
+
+static xccl_status_t
+xccl_mrail_collective_init(xccl_coll_op_args_t *coll_args,
+                           xccl_tl_coll_req_t **request, xccl_tl_team_t *tl_team)
+{
+    xccl_mrail_coll_req_t *req;
+    xccl_mrail_team_t     *team = ucs_derived_of(tl_team, xccl_mrail_team_t);
+    xccl_status_t         status;
+    xccl_coll_op_args_t   team_args;
+    int                   n_reqs;
+    int                   buf_len;
+    int                   last_buf_len;
+    int                   len;
+    xccl_mrail_debug("Entering collective init");
+    req = (xccl_mrail_coll_req_t*)malloc(sizeof(xccl_mrail_coll_req_t));
+    
+    if (req == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+
+    if (coll_args->buffer_info.len < team->n_teams) {
+        n_reqs  = 1;
+        buf_len = 0;
+    }
+    else {
+        n_reqs  = team->n_teams;
+        buf_len = coll_args->buffer_info.len / n_reqs;
+    }
+    last_buf_len = coll_args->buffer_info.len - (n_reqs - 1)*buf_len;
+
+    memset(req->reqs, 0, n_reqs * sizeof(xccl_coll_req_h));
+    team_args = *coll_args;
+
+    for(req->n_reqs = 0; req->n_reqs < n_reqs; req->n_reqs++) {
+        len = buf_len;
+        if (req->n_reqs == n_reqs - 1) {
+            len = last_buf_len;
+        }
+        xccl_mrail_debug("len %d \n", len);
+        team_args.buffer_info.dst_buffer = coll_args->buffer_info.dst_buffer +
```

Comment:
ptr arithmetics on void * is a bad practice (https://stackoverflow.com/questions/3523145/pointer-arithmetic-for-void-pointer-in-c). Please use smth like:
(void*)((ptrdiff_t)coll_args->buffer_info.dst_buffer + len*req->nreqs)